### PR TITLE
[Merged by Bors] - chore(Data/Set): split Finite.lean into `Defs`, `Basic`, `Lemmas`

### PR DESCRIPTION
**Pull request by Vierkantor**
`Data.Set.Finite` had a lot of imports and in turn was imported often. So let's split it up along import lines:
 * `Data/Finite/Defs.lean` now contains the definition of `Set.Finite`
 * `Data/Set/Finite/Basic.lean` knows only `Set.Finite` and `Finset`
 * `Data/Finite/Prod.lean` now contains finiteness results of products of sets
 * `Data/Fintype/Pi.lean` now contains finiteness results of sets of functions
 * `Data/Set/Finite/Lattice.lean` contains finiteness results of lattice operations on sets
 * `Data/Set/Finite/List.lean` contains finiteness results on lists converted to sets
 * `Data/Set/Finite/Monad.lean` also knows about the monad structure on `Set`
 * `Data/Set/Finite/Powerset.lean` contains finiteness results on powersets
 * `Data/Set/Finite/Range.lean` contains finiteness results on `Set.range`
 * `Data/Set/Finite/Lemmas.lean` is for lemmas *about* `Set.Finite` (and therefore does not contain any `Finite` instances itself).

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by YaelDillies** [comment]
Hmm, Bhavik and I had another plan. I will write more later

**Comment by Ruben-VandeVelde** [comment]
@YaelDillies please do, when you're able:)

**Comment by YaelDillies** [comment]
Sorry for the delay. My wrists feel better today, so here are my thoughts:
1. `Finite` and `Set.Finite` are really close
2. To fight API getting locally linearised over time, we want to ensure that a local API (eg the `Set.Finite` API) has as many non-linearisable leaf files as possible.
3. "Construction files" (eg files about `Set.Finite` and `Prod`/`Pi`/`Option`, ...) are a perfect example of non-linearisable files (very few interactions exist/are stated between the various possible constructions).

Therefore, Bhavik and I's plan was to/I would suggest you do the following:
1. Put `Finite` and `Set.Finite` in the same file, to make sure the latter doesn't get considered heavier than the former. Concretely, merge `Data.Finite.Defs` and `Data.Set.Finite.Defs`.
2. Make `Data.Finite.<Construction>` a leaf file locally. Concretely, swap the imports between `Data.Finite.<Construction>` and `Data.Set.Finite.Lemmas`. Hopefully, `Data.Set.Finite.Lemmas` doesn't need existing anymore.

**Comment by Vierkantor** [comment]
Thanks for writing out your suggestions, I'll try them out and see how it goes.

**Comment by Vierkantor** [comment]
Okay, everything builds so I think this is ready for review again. Does this match your plans somewhat, @YaelDillies? :)

**Comment by YaelDillies** [comment]
Something seems off. Why is this PR adding 600 more lines than it is removing?

File: Mathlib/Data/Set/CoeSort.lean
Comment (YaelDillies) [inline Mathlib/Data/Set/CoeSort.lean:1]: Why are you splitting this off `Data.Set.Operations`? Should this be a separate PR?

File: Mathlib/Data/Set/Finite/Pi.lean
Comment (YaelDillies) [inline Mathlib/Data/Set/Finite/Pi.lean:1]: This could merge with `Data.Fintype.Pi` (where the `Finite` instance already is) at little cost

File: Mathlib/Data/Set/Finite/Prod.lean
Comment (YaelDillies) [inline Mathlib/Data/Set/Finite/Prod.lean:1]: This could merge with `Data.Fintype.Prod` (where the `Finite` instance already is) at little cost

**Review by YaelDillies — COMMENTED** [review]
Having gone through the entire diff, it looks like those 600 new lines are all accounted for by the module docs of the new files! 😳

I like the new file structure, but I would still like `Finite` and `Set.Finite` to be closer to each other. See below for concrete suggestions.

File: Mathlib/Data/Set/CoeSort.lean
Comment (Vierkantor) [reply Mathlib/Data/Set/CoeSort.lean:1]: `Data.Set.Operations` pulls in quite a bit of unneeded imports via `Logic.Function.Basic` and `Mathlib.Data.SProd`. But looking at the import graph right now, it doesn't seem quite as hefty as I remembered, so I'm not opposed to dropping the change, or adding it to a separate PR for further debate. What would you prefer?

File: Mathlib/Data/Set/Finite/Pi.lean
Comment (Vierkantor) [reply Mathlib/Data/Set/Finite/Pi.lean:1]: I don't see any `Finite` instances in that file. In fact, `Pi.finite` lives in `Data.Finite.Prod`!

File: Mathlib/Data/Set/Finite/Prod.lean
Comment (Vierkantor) [reply Mathlib/Data/Set/Finite/Prod.lean:1]: I assume you mean merging with `Data.Finite.Prod`?

Comment (YaelDillies) [reply Mathlib/Data/Set/Finite/Prod.lean:1]: I meant fintype because I thought the `Finite` instance was there as well, but it seems not. So `Data.Finite.Prod` is good

File: Mathlib/Data/Set/CoeSort.lean
Comment (YaelDillies) [reply Mathlib/Data/Set/CoeSort.lean:1]: I thought `Data.Set.Operations` was supposed to be very light, so I am wary of atomising it even further. Either dropping the change or moving it to a separate PR would be fine

Comment (Vierkantor) [reply Mathlib/Data/Set/CoeSort.lean:1]: I just dropped the change.

File: Mathlib/Data/Set/Finite/Pi.lean
Comment (YaelDillies) [reply Mathlib/Data/Set/Finite/Pi.lean:1]: Oh!

File: Mathlib/Data/Set/CoeSort.lean
Comment (Vierkantor) [reply Mathlib/Data/Set/CoeSort.lean:1]: Oh, now there's a build error in `Mathlib/Algebra/Group/Int.lean:20:18`: Declaration `@Set.range` is not allowed to be imported by this file.
```
It is defined in Mathlib.Data.Set.Operations,
  which is imported by Mathlib.Data.Finite.Defs,
  which is imported by Mathlib.Algebra.Group.TypeTags,
  which is imported by Mathlib.Algebra.Group.Even,
  which is imported by Mathlib.Algebra.Group.Nat,
  which is imported by this file.
```

File: Mathlib/Data/Set/Finite/Pi.lean
Comment (Vierkantor) [reply Mathlib/Data/Set/Finite/Pi.lean:1]: Actually, I think `Fintype.Pi` is a good idea since that file would import `Finite.Prod`. So let me do that...

Comment (Vierkantor) [reply Mathlib/Data/Set/Finite/Pi.lean:1]: You were right all along! `Fintype.Pi` is the right place to put these :D

File: Mathlib/Data/Set/CoeSort.lean
Comment (Vierkantor) [reply Mathlib/Data/Set/CoeSort.lean:1]: I have reverted the revert of the splitting, and I'll make a new PR for it. (Which this one will need to depend on.)

Comment (Vierkantor) [reply Mathlib/Data/Set/CoeSort.lean:1]: True, and I don't like the dependency on the other PR. I think I will drop the `assert_not_exists` and make a note to readd it in #19031.

**Comment by YaelDillies** [comment]
Thanks! This looks very good now. The extra 10 imports to `Data.Fintype.Pi` are actually just very basic tactic files, while the -60 imports to some files is substantial, so this seems to be a net win.

maintainer merge

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge
